src/lsm/segmented_array.zig

```diff
@@ -219,98 +285,83 @@ fn SegmentedArrayType(
-            //
-            // elements: [yy], relative_index: 2
-            // [xxx_]
-            // [xxy_][yx__]
-            //
-            // elements: [yy], relative_index: 5
-            // [xxxxxx]
-            // [xxxx__][xyyx__]
-            //
-            // elements: [yyyyy_], relative_index: 5
-            // [xxxxx_]
-            // [xxxxx_][yyyyy_]
-
-            assert(cursor.relative_index <= array.count(a));
-
-            const existing_a_head = a_pointer[0..@minimum(a_half, cursor.relative_index)];
-            const existing_b_head = a_pointer[existing_a_head.len..cursor.relative_index];
-
-            const existing_a_tail = a_pointer[existing_a_head.len..][0..a_half -|
-                (cursor.relative_index + elements.len)];
-            const existing_b_tail = a_pointer[existing_a_head.len + existing_a_tail.len +
-                existing_b_head.len .. array.count(a)];
-
-            const elements_a = elements[0 .. a_half - existing_a_head.len - existing_a_tail.len];
-            const elements_b = elements[elements_a.len..];
-
-            assert(array.count(a) == existing_a_head.len + existing_b_head.len +
-                existing_a_tail.len + existing_b_tail.len);
-            assert(elements.len == elements_a.len + elements_b.len);
-            assert(a_half == existing_a_head.len + elements_a.len + existing_a_tail.len);
-            assert(b_half == existing_b_head.len + elements_b.len + existing_b_tail.len);
-            assert(total == a_half + b_half);
-
-            if (existing_a_tail.len > 0) assert(existing_b_head.len == 0 and elements_b.len == 0);
-            if (existing_b_head.len > 0) assert(elements_a.len == 0 and existing_a_tail.len == 0);
-            if (elements_a.len > 0) assert(existing_b_head.len == 0);
-            if (elements_b.len > 0) assert(existing_a_tail.len == 0);
-
-            assert(existing_a_head.ptr == a_pointer);
-            assert(existing_a_head.ptr + existing_a_head.len == existing_a_tail.ptr);
-            assert(existing_a_head.ptr + existing_a_head.len == existing_b_head.ptr);
-            if (existing_b_head.len > 0) {
-                assert(existing_b_head.ptr + existing_b_head.len == existing_b_tail.ptr);
-            } else {
-                assert(existing_a_tail.ptr + existing_a_tail.len == existing_b_tail.ptr);
-            }
+            const a_half_pointer = a_pointer[0..a_half];
+            const b_half_pointer = b_pointer[0..b_half];
 
-            mem.copy(T, b_pointer[existing_b_head.len + elements_b.len ..], existing_b_tail);
-            mem.copy(T, b_pointer[existing_b_head.len..], elements_b);
-            mem.copy(T, b_pointer, existing_b_head);
+            // Move part of `a` forwards to make space for elements.
+            copy_backwards(
+                a_half_pointer,
+                b_half_pointer,
+                cursor.relative_index + elements.len,
+                a_pointer[cursor.relative_index..a_count],
+            );
```

Comment:
I prefer the assertions and inline copies as is, i.e. lots of upfront assertions and then three simple memcopies, without the `copy_backwards()` indirection. The idea is to contain all the complexity in the same scope, with a minimum of cached variables.

What are our motivations for refactoring this?

Reply:
> three simple memcopies

5 memcopies that I see.

This seems like the natural way to bring the size of this function down — previously ~90 lines not counting the comments.

Reply:
My thinking was that the actual logic is simple: move everything after the insertion point along by elements.len, and then copy elements into the gap. It's only made complicated by the fact that the array is actually two arrays and the logic to deal with that has been inlined into each of the copies. Un-inlining that logic makes the proof simpler.

But I'm not that invested in the change. It was just a side-effect of sketching out a proof that the original code was correct.

Reply:
Thanks for the explanation @jamii — are the number of actual copies done always the same across both versions for all the cases? If so, then let's go with your refactor.

Reply:
@jorangreef I restored the original comment, and expanded it to show the step-by-step behavior.

Reply:
> are the number of actual copies done always the same across both versions for all the cases

Yeah, in both cases they do max 5 memcopies and each element is copied at most once.